src/include/zarr.h

```diff
@@ -0,0 +1,550 @@
+      const ZarrStreamSettings* settings,
+      size_t index,
+      char* name,
+      size_t bytes_of_name,
+      ZarrDimensionType* kind,
+      size_t* array_size_px,
+      size_t* chunk_size_px,
+      size_t* shard_size_chunks);
+
+    uint8_t ZarrStreamSettings_get_multiscale(
+      const ZarrStreamSettings* settings);
+
+    /***************************************************************************
+     * Zarr stream
+     *
+     * The Zarr stream struct is created using ZarrStream_create and destroyed
+     * using ZarrStream_destroy. If the struct fails to be created,
+     * ZarrStream_create returns NULL.
+     **************************************************************************/
+
+    typedef enum
+    {
+        ZarrVersion_2 = 2,
+        ZarrVersion_3,
+        ZarrVersionCount
+    } ZarrVersion;
+
+    typedef struct ZarrStream_s ZarrStream;
+
+    /**
+     * @brief Create a Zarr stream.
+     * @param[in, out] settings The settings for the Zarr stream. This pointer
+     * is consumed by this function.
+     * @param[in] version The version of the Zarr stream. 2 or 3.
+     * @return A pointer to the Zarr stream struct, or NULL on failure.
+     */
+    ZarrStream* ZarrStream_create(ZarrStreamSettings* settings,
+                                  ZarrVersion version);
+    void ZarrStream_destroy(ZarrStream* stream);
+
+    /***************************************************************************
+     * Writing data to the Zarr stream.
+     *
+     * This function writes data to the Zarr stream. It returns a ZarrError to
+     * indicate success or failure, which can be converted to a human-readable
+     * error message using Zarr_get_error_message().
+     **************************************************************************/
+
+    /**
+     * @brief Append data to the Zarr stream.
+     * @param[in, out] stream The Zarr stream struct.
+     * @param[in] data The data to append.
+     * @param[in] bytes_in The number of bytes in @p data. It should be at least
+     * the size of a single frame.
+     * @param[out] bytes_out The number of bytes written to the stream.
+     * @return ZarrError_Success on success, or an error code on failure.
+     */
+    ZarrError ZarrStream_append(ZarrStream* stream,
+                                const void* data,
+                                size_t bytes_in,
+                                size_t* bytes_out);
+
+    /***************************************************************************
+     * Functions for getting parameters on the Zarr stream.
+     *
+     * These functions return the value of the specified parameter.
+     * If the Zarr stream is NULL, the functions return NULL or 0.
+     **************************************************************************/
+
+    ZarrVersion ZarrStream_get_version(const ZarrStream* stream);
+
+    const char* ZarrStream_get_store_path(const ZarrStream* stream);
```

Comment:
Are these getters for the same underlying object, but instead of passing the settings object, you passed the stream? If that's the case, why not use one or the other? I'm still trying to understand the benefit of exposing every parameter as an accessor method, so I'm surprised to see another set of accessor functions.

I may need to learn more about this API, but currently it feels like a lot of unnecessary flexibility (and maybe a violation of the principle of least exposure). I might be misunderstanding the library's purpose, but I'm wondering: if I've already provided this information, why are accessor methods needed to retrieve it? Wouldn't returning an instance of the settings object be more straightforward?